drivers/axi_core/axi_dmac/axi_dmac.c

```diff
@@ -123,204 +232,192 @@ int32_t axi_dmac_write(struct axi_dmac *dmac,
+			break;
+		default:
+			return FAILURE;
+		}
 
-	if ((size - 1) > dmac->transfer_max_size) {
-		dmac->big_transfer.address = address;
-		dmac->big_transfer.size = size - 1;
-		dmac->big_transfer.size_done = dmac->transfer_max_size;
+		/* Compute remaining size. */
+		dmac->remaining_size = dmac->remaining_size - (burst_size + 1);
 
-		axi_dmac_write(dmac, AXI_DMAC_REG_X_LENGTH,
-			       dmac->transfer_max_size);
+		/* Specify the length of the transfer and trigger transfer. */
+		axi_dmac_write(dmac, AXI_DMAC_REG_X_LENGTH, burst_size);
 		axi_dmac_write(dmac, AXI_DMAC_REG_Y_LENGTH, 0x0);
-
 		axi_dmac_write(dmac, AXI_DMAC_REG_FLAGS, dmac->flags);
-
 		axi_dmac_write(dmac, AXI_DMAC_REG_START_TRANSFER, 0x1);
-
-		while(!dmac->big_transfer.transfer_done) {
-			timeout++;
-			if (timeout == UINT32_MAX)
-				return -1;
-		}
-
-		dmac->big_transfer.address = 0;
-		dmac->big_transfer.size = 0;
-		dmac->big_transfer.size_done = 0;
-
-		return 0;
+	} else {
+		return FAILURE;
 	}
 
-	axi_dmac_write(dmac, AXI_DMAC_REG_X_LENGTH, size - 1);
-	axi_dmac_write(dmac, AXI_DMAC_REG_Y_LENGTH, 0x0);
-
-	axi_dmac_write(dmac, AXI_DMAC_REG_FLAGS, dmac->flags);
-
-	axi_dmac_write(dmac, AXI_DMAC_REG_START_TRANSFER, 0x1);
-
-	if (dmac->flags & DMA_CYCLIC)
-		return 0;
-
-	/* Wait until the new transfer is queued. */
-	do {
-		axi_dmac_read(dmac, AXI_DMAC_REG_START_TRANSFER, &reg_val);
-	} while(reg_val == 1);
-
-	/* Wait until the current transfer is completed. */
-	do {
-		axi_dmac_read(dmac, AXI_DMAC_REG_IRQ_PENDING, &reg_val);
-		if (reg_val == (AXI_DMAC_IRQ_SOT | AXI_DMAC_IRQ_EOT))
-			break;
-	} while(!dmac->big_transfer.transfer_done);
-	if (reg_val != (AXI_DMAC_IRQ_SOT | AXI_DMAC_IRQ_EOT))
-		axi_dmac_write(dmac, AXI_DMAC_REG_IRQ_PENDING, reg_val);
-
-	/* Wait until the transfer with the ID transfer_id is completed. */
-	do {
-		axi_dmac_read(dmac, AXI_DMAC_REG_TRANSFER_DONE, &reg_val);
-	} while((reg_val & (1u << transfer_id)) != (1u << transfer_id));
-
 	return 0;
 }
 
-/***************************************************************************//**
- * @brief axi_dmac_init
+/*******************************************************************************
+ * @brief wait for DMA transfer to be completed
  *******************************************************************************/
-int32_t axi_dmac_init(struct axi_dmac **dmac_core,
-		      const struct axi_dmac_init *init)
+int32_t axi_dmac_transfer_wait_completion(struct axi_dmac *dmac)
 {
-	struct axi_dmac *dmac;
-
-	dmac = (struct axi_dmac *)calloc(1, sizeof(*dmac));
-	if (!dmac)
-		return -1;
-
-	dmac->name = init->name;
-	dmac->base = init->base;
-	dmac->direction = init->direction;
-	dmac->flags = init->flags;
-	dmac->transfer_max_size = -1;
-
-	axi_dmac_write(dmac, AXI_DMAC_REG_X_LENGTH, dmac->transfer_max_size);
-	axi_dmac_read(dmac, AXI_DMAC_REG_X_LENGTH, &dmac->transfer_max_size);
+	uint32_t timeout = 0;
 
-	*dmac_core = dmac;
+	while(!dmac->transfer.transfer_done) {
+		timeout++;
```

Comment:
nitpick: we could use the `mdelay` API to implement a more deterministic timeout.

Reply:
Solved in subsequent commit.